mindmeld/models/embedder_models.py

```diff
@@ -0,0 +1,184 @@
+        """
+        self.model_name = kwargs.get("model_name", "default")
+        self.cache_path = path.get_embedder_cache_file_path(
+            app_path, kwargs.get("embedder_type", "default"), self.model_name
+        )
+
+        folder = os.path.dirname(self.cache_path)
+        if not os.path.isdir(folder):
+            os.makedirs(folder)
+
+        if not os.path.exists(self.cache_path) or os.path.getsize(self.cache_path) == 0:
+            self.cache = {}
+        else:
+            with open(self.cache_path, "rb") as fp:
+                self.cache = pickle.load(fp)
+
+        self.model = self.load(**kwargs)
+
+    @abstractmethod
+    def load(self, **kwargs):
+        """Loads the embedder model
+
+        Returns:
+            The model object.
+        """
+        pass
+
+    @abstractmethod
+    def encode(self, text_list):
+        """
+        Args:
+            text_list (list): A list of text strings for which to generate the embeddings.
+
+        Returns:
+            (list): A list of numpy arrays of the embeddings.
+        """
+        pass
+
+    def clear_cache(self):
+        """Deletes the cache file.
+        """
+        if os.path.exists(self.cache_path):
+            os.remove(self.cache_path)
+
+    def get_encodings(self, text_list):
+        """Fetches the encoded values from the cache, or generates them.
+        Args:
+            text_list (list): A list of text strings for which to get the embeddings.
+
+        Returns:
+            (list): A list of numpy arrays with the embeddings.
+        """
+        encoded = [self.cache.get(text, None) for text in text_list]
+        cache_miss_indices = [i for i, vec in enumerate(encoded) if vec is None]
+        text_to_encode = [text_list[i] for i in cache_miss_indices]
+        model_encoded_text = self.encode(text_to_encode)
+
+        for i, v in enumerate(cache_miss_indices):
+            encoded[v] = model_encoded_text[i]
+            self.cache[text_to_encode[i]] = model_encoded_text[i]
+        return encoded
+
+    def dump(self):
+        """Dumps the cache to disk.
+        """
+        with open(self.cache_path, "wb") as fp:
+            pickle.dump(self.cache, fp)
+
+
+class BertEmbedder(Embedder):
+    """
+    Encoder class for bert models as described here: https://github.com/UKPLab/sentence-transformers
+    """
+
+    def load(self, **kwargs):
+        DEFAULT_BERT = "bert-base-nli-mean-tokens"
+        if self.model_name == "default":
+            bert_model_name = DEFAULT_BERT
+            logger.info("No bert model specifications passed, using default.")
+        else:
+            bert_model_name = self.model_name
+        return SentenceTransformer(bert_model_name)
+
+    def encode(self, text_list):
+        return self.model.encode(text_list)
+
+
+def get_query_tokens(query):
+    """Splits into tokens, removes punctuation, and removes whitespace.
+    """
+    query = query.translate(str.maketrans("", "", string.punctuation))
```

Comment:
I am not exactly sure what the `str.maketrans` is doing:
```
>>> str.maketrans("", "", string.punctuation)
{33: None, 34: None, 35: None, 36: None, 37: None, 38: None, 39: None, 40: None, 41: None, 42: None, 43: None, 44: None, 45: None, 46: None, 47: None, 58: None, 59: None, 60: None, 61: None, 62: None, 63: None, 64: None, 91: None, 92: None, 93: None, 94: None, 95: None, 96: None, 123: None, 124: None, 125: None, 126: None}
```

Reply:
This line removes punctuation.

https://docs.python.org/3.3/library/stdtypes.html?highlight=maketrans#str.maketrans

So with maketrans, the first two arguments are mapping empty to empty (no-op), the third argument is mapping all punctuation to None. This passed to query.translate, will give us the query with punctuation removed.

Reply:
Should we use our own tokenizer for this, or is this meant to align with BERT preprocessing? Like this would squash "language-dependent", "smeagol@precious-my.webex.com", "absolutely—that" into single tokens.

Reply:
I think the two are similar enough, updated to use the MM tokenizer.